### PR TITLE
🐛 Avoid crash when formstate objects change shape

### PR DESCRIPTION
**Pull request by dmnd**
A crash was reported where a formstate object changes shape after a
`customChange`. The new shape was being validated by a collection of
validation functions arranged in the previous object shape, causing all
sorts of mayhem.

To avoid this, remove all validation functions after `customChange`,
then render, causing new validation functions to be added again (in the
correct shape). Only after the new shape has been built up does
validation run.

Test plan:

* @larajanecka & I verified the steps that crashed bid app before this change are now working.
* The new tests crash prior to the changes made in this PR:

  ```bash
  git checkout origin/master
  git checkout custom-change-validation-fn-rebuild -- src/test/ArrayField.test.js src/test/ObjectField.test.js
  yarn test
  ```

### CODE REVIEW

File: src/Form.js [hunk 1]
```diff
@@ -450,7 +462,12 @@ export default class Form<T, ExtraSubmitData> extends React.Component<
         extras => ({...extras, errors: {...extras.errors, client: errors}}),
         tree
       );
-      return {formState: [rootValue, updatedTree]};
+
+      const formState = [rootValue, updatedTree];
+
+      this.props.onValidation(isValid(formState));
```
Comment (zgotsch) [inline src/Form.js]: A couple things to mention here:

1. I don't believe `isValid` looks at feedback strategy. We should make that clear and probably establish some API so that you can get that info also.
2. It may be surprising to users that `onValidation` is called on `blur`, since validations are not run then (it's surprising to me!)
3. If this blur changes whether errors are being shown, how should we communicate that via the API?

Comment (dmnd) [reply src/Form.js]: This isn't `_handleBlur`, that's just github collapsing the file. This code is inside of `recomputeErrorsAtPathAndRender`. But your point still stands because that gets called all over the place 😄 

Before this PR is out of draft state I wanted to think through if it actually made sense to put here. For now I added it here because of the branch at https://github.com/flexport/formula-one/pull/57/files#diff-03433190399c5167b022941d6db4acebR430

**Comment by dmnd** [comment]
I manually tested this against the crash that @larajanecka reported and it avoid the crash (I think). I'll confirm that tomorrow in person, but in the meantime @zgotsch this PR is good to review.

**Comment by dmnd** [comment]
I confirmed with @larajanecka that this PR fixes the issue in bid app.

File: src/Form.js (hunk 1 above)
Comment (dmnd) [reply src/Form.js]: For posterity: we thought it through and decided that it does make sense to call `onValidation` here, and the previous implementation was buggy by not doing so.

File: src/Form.js [hunk 2]
```diff
@@ -555,8 +576,18 @@ export default class Form<T, ExtraSubmitData> extends React.Component<
...
-            applyCustomChangeToTree(path, formState, this.validations),
+          applyCustomChangeToTree: (path, [value, _tree]) => {
+            this.validations = removeChildValidationsAtPath(
+              path,
+              this.validations
+            );
+            invariant(
+              this.pathForPostCommitValidation === null,
+              "Unexpected pending validation. This is a bug in Formula One, please report it."
+            );
```
Comment (dmnd) [inline src/Form.js]: Write a comment here that explains why this is not a queue for now.

TL;DR: The formState could transition through multiple unvalidated shapes and generally fuck up the validations map, so it's unclear what should happen

File: src/test/ObjectField.test.js [hunk 3]
```diff
@@ -541,5 +541,51 @@ describe("ObjectField", () => {
...
+          </Form>
+        );
+      }
+
+      const renderer = TestRenderer.create(<App />);
+      const inner = renderer.root.findAllByType(TestInput)[0];
+      expect(() => {
+        inner.instance.change("b");
+      }).not.toThrow();
+    });
```
Comment (dmnd) [inline src/test/ObjectField.test.js]: Write the `ArrayField` version of this test too!

File: src/Form.js (hunk 1 above)
Comment (dmnd) [reply src/Form.js]: Filed #59 and #60 to track the other concerns here.

File: src/Form.js [hunk 4]
```diff
@@ -555,8 +576,18 @@ export default class Form<T, ExtraSubmitData> extends React.Component<
         value={{
           shouldShowError: this.props.feedbackStrategy.bind(null, metaForm),
           registerValidation: this.handleRegisterValidation,
-          applyCustomChangeToTree: (path, formState) =>
-            applyCustomChangeToTree(path, formState, this.validations),
+          applyCustomChangeToTree: (path, [value, _tree]) => {
+            this.validations = removeChildValidationsAtPath(
```
[the remark this answers is not in the document]
Comment (dmnd) [reply src/Form.js]: Renamed the function, leaving the more general task for a different PR, because most of the `AtPath` symbols are referring to nodes (not trees).

File: src/test/ObjectField.test.js (hunk 3 above)
Comment (dmnd) [reply src/test/ObjectField.test.js]: Done, see ecd2a40

**Comment by dmnd** [comment]
@zgotsch please take another look!
